File: coderust/arrays/optimal_job_selection.rb
def optimal_job_selection(some_jobs)
  sorted_jobs = some_jobs.sort_by {|jobs| jobs[1]}

  current_job = sorted_jobs.first
  selected_jobs = [current_job]
  i = 1

  while i < sorted_jobs.length
    if sorted_jobs[i][0] > current_job[1]
      selected_jobs.push(sorted_jobs[i])
    end

    i += 1
  end

  selected_jobs
end

intervals = [[11, 14], [3, 12], [1, 4], [7, 10], [4, 10]]

p optimal_job_selection(intervals)

File: coderust/arrays/optimal_job_selection.py
def optimal_job_selection(sample_jobs):
    sorted_intervals = sorted(sample_jobs, key=lambda x:x[1])
    initial_job_start, initial_job_end = sorted_intervals[0]
    selected_jobs = [(initial_job_start, initial_job_end)]

    for current_job_start, current_job_end in sorted_intervals[1:]:
        if current_job_start > initial_job_end:
            selected_jobs.append((current_job_start, current_job_end))

    return selected_jobs

intervals = [(11, 14), (3, 12), (1, 4), (7, 10), (4, 10)]
print optimal_job_selection(intervals)
